Dockerfile

```diff
@@ -0,0 +1,179 @@
+        echo 'opcache.enable_cli = 1'; \
+        echo 'opcache.memory_consumption = 256'; \
+        echo 'opcache.interned_strings_buffer = 16'; \
+        echo 'opcache.max_accelerated_files = 20011'; \
+        echo 'opcache.fast_shutdown = 1'; \
+        echo 'realpath_cache_size = 4096K'; \
+        echo 'realpath_cache_ttl = 600'; \
+    } > /usr/local/etc/php/php.ini
+
+RUN { \
+        echo 'date.timezone = UTC'; \
+        echo 'short_open_tag = off'; \
+        echo 'memory_limit = -1'; \
+    } > /usr/local/etc/php/php-cli.ini
+
+CMD ["php-fpm"]
+
+
+#####################################
+##             APP DEV             ##
+#####################################
+FROM app as app-dev
+
+ARG NODE_VERSION
+ARG COMPOSER_VERSION
+ARG XDEBUG_VERSION
+
+ENV COMPOSER_ALLOW_SUPERUSER=1
+ENV APP_ENV=dev
+
+# Install paquet requirements
+RUN set -ex; \
+    # Install required system packages
+    apt-get update; \
+    apt-get install -qy --no-install-recommends \
+            unzip \
+            git \
+    ; \
+    # Clean aptitude cache and tmp directory
+    apt-get clean && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*;
+
+# Install Node
+RUN set -ex; \
+    curl -L -o /tmp/nodejs.tar.gz https://nodejs.org/dist/v${NODE_VERSION}/node-v${NODE_VERSION}-linux-x64.tar.gz; \
+    tar xfvz /tmp/nodejs.tar.gz -C /usr/local --strip-components=1; \
+    rm -f /tmp/nodejs.tar.gz; \
+    npm install yarn -g
+
+# Install Composer
+RUN set -ex; \
+    EXPECTED_SIGNATURE="$(curl -L https://getcomposer.org/download/${COMPOSER_VERSION}/composer.phar.sha256sum)"; \
+    curl -L -o composer.phar https://getcomposer.org/download/${COMPOSER_VERSION}/composer.phar; \
+    ACTUAL_SIGNATURE="$(sha256sum composer.phar)"; \
+    if [ "$EXPECTED_SIGNATURE" != "$ACTUAL_SIGNATURE" ]; then >&2 echo 'ERROR: Invalid installer signature' && rm /usr/local/bin/composer && exit 1 ; fi; \
+    chmod +x composer.phar && mv composer.phar /usr/local/bin/composer; \
+    RESULT=$?; \
+    exit $RESULT;
+
+# Edit OPCache configuration
+RUN set -ex; \
+    { \
+        echo 'opcache.validate_timestamps = 1'; \
+        echo 'opcache.revalidate_freq = 0'; \
+    } >> /usr/local/etc/php/php.ini
+
+# Install Xdebug
+RUN set -ex; \
+    if [ "${XDEBUG_VERSION}" != 0 ]; \
+    then \
+        pecl install xdebug-${XDEBUG_VERSION}; \
+        docker-php-ext-enable xdebug; \
+        { \
+            echo 'xdebug.remote_enable = on'; \
+            echo 'xdebug.remote_connect_back = on'; \
+        } >> /usr/local/etc/php/php.ini \
+    ; fi
+
+
+#####################################
+##       PROD ASSETS BUILDER       ##
+#####################################
+FROM node:${NODE_VERSION} as assets-builder
+
+COPY . /app
+WORKDIR /app
+
+RUN yarn install && yarn build && rm -R node_modules
+
+#####################################
+##       PROD VENDOR BUILDER       ##
+#####################################
+FROM composer:${COMPOSER_VERSION} as vendor-builder
```

Comment:
Here I'd inherit from `app` step, because if application relies on some extension installed before (in app step) (example bcmath) here this dependency will be missing. Proposal:

```
FROM app as vendor-builder

ARG COMPOSER_VERSION

# Install Composer
RUN set -ex; \
    EXPECTED_SIGNATURE="$(curl -L https://getcomposer.org/download/${COMPOSER_VERSION}/composer.phar.sha256sum)"; \
    curl -L -o composer.phar https://getcomposer.org/download/${COMPOSER_VERSION}/composer.phar; \
    ACTUAL_SIGNATURE="$(sha256sum composer.phar)"; \
    if [ "$EXPECTED_SIGNATURE" != "$ACTUAL_SIGNATURE" ]; then >&2 echo 'ERROR: Invalid installer signature' && rm /usr/local/bin/composer && exit 1 ; fi; \
    chmod +x composer.phar && mv composer.phar /usr/local/bin/composer; \
    RESULT=$?; \
    exit $RESULT;

```

Reply:
Previously, I was doing like it, but I've change to avoid have composer in the final image.

At the moment on all my projects, I've never had this problem (but, sure, I've not test all dependencies). The goal of this part is to only install vendors and avoid to have composer in the final image.

Reply:
Final image won't have this dependency, because `app-prod` inherits from `app`, not `vendor-builder`. You wisely used only copy of `/app` files from `vendor-builder` while composer is located under `/usr/local/bin/composer`.

Reply:
Yes but if we install composer on `app` then, you have `composer` on `app-prod` and finaly on the image.

Reply:
Yes yes, that's why my proposal is to have it only on `vendor-builder`:

```
FROM app as vendor-builder
```